ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp

```diff
@@ -435,188 +632,368 @@ string get_login(string username, string password, string cap)
   return gs_str_submitted;
 }
 
-string get_string(string message,string def,string cap)
+string get_string(string str, string def, string title)
 {
-  gs_cap = cap; gs_message = message; gs_def = def;
-  DialogBox(enigma::hInstance,"getstringdialog",enigma::hWnd,GetStrProc);
+  tstring tstrStr = widen(str);
+  tstring tstrDef = widen(def);
+  if (title == "") title = window_get_caption();
+  gs_cap = title;
 
-  return gs_str_submitted;
+  wcsncpy(wstrPromptStr, tstrStr.c_str(), 4096);
+  wcsncpy(wstrTextEntry, tstrDef.c_str(), MAX_PATH);
+
+  HideInput = 0;
+  DialogBoxW(enigma::hInstance, MAKEINTRESOURCEW(IDD_INPUTBOX), enigma::hWnd, reinterpret_cast<DLGPROC>(InputBoxHookProc));
+
+  static string strResult;
+  strResult = shorten(wstrTextEntry);
+  return strResult;
 }
 
-int get_integer(string message,string def,string cap)
+string get_password(string str, string def, string title)
 {
-  gs_cap = cap; gs_message = message; gs_def = def;
-  DialogBox(enigma::hInstance,"getstringdialog",enigma::hWnd,GetStrProc);
-  if (gs_str_submitted == "") return 0;
-  puts(gs_str_submitted.c_str());
+  tstring tstrStr = widen(str);
+  tstring tstrDef = widen(def);
+  if (title == "") title = window_get_caption();
+  gs_cap = title;
+
+  wcsncpy(wstrPromptStr, tstrStr.c_str(), 4096);
+  wcsncpy(wstrTextEntry, tstrDef.c_str(), MAX_PATH);
 
-  return atol(gs_str_submitted.c_str());
+  HideInput = 1;
+  DialogBoxW(enigma::hInstance, MAKEINTRESOURCEW(IDD_INPUTBOX), enigma::hWnd, reinterpret_cast<DLGPROC>(InputBoxHookProc));
+
+  static string strResult;
+  strResult = shorten(wstrTextEntry);
+  return strResult;
 }
 
-double get_number(string message,string def,string cap)
+double get_integer(string str, double def, string title)
```

Comment:
I think this more of a bug with GMA. It shouldn't be able to take a string. The whole point of the function is to input a double because that is the only thing it can possibly return, just like get_string can only return a string

Reply:
I mean this I'm not really sure about tbh and do think Josh should decide on this particular one.

Reply:
@JoshDreamland what do you want me to do with this one?

Reply:
I agree with you, but it would be nice if we could valid the input. For example, block the text control from accepting non-numeric characters. Win32 should have a flag for it.

It looks like `ES_NUMBER` is a thing, we should perhaps use it. Also, it would be good to verify that flag accepts floats and doubles with a decimal. The other issue is numbers in a different format, such as currency or hexadecimal. I think this is why GMS just strips the non-digit characters off the string and gives it to you as-is.
http://www.cplusplus.com/forum/windows/107187/